### PR TITLE
[REF] web, *: move progress bar to kanban view

### PR DESCRIPTION
**Pull request by jpp-odoo**
The aim of this commit is to remove the specific complexity of the
kanban's progress bar from the relational model, and put it into the
kanban view itself.

This commit prepares and is part of the task that aim to refactor the
relational model and migrate it to owl.

part-of task-id 3179751

Co-authored-by: Aaron Bohy <aab@odoo.com>

### CODE REVIEW

File: addons/web/static/src/views/kanban/kanban_controller.js [hunk 1]
```diff
@@ -93,10 +108,23 @@ export class KanbanController extends Component {
...
+        if (!progressAttributes) {
+            return [];
+        }
+        if (progressAttributes.sumField) {
+            return [progressAttributes.sumField];
+        } else {
+            return [];
+        }
+    }
+
```
Comment (jpp-odoo) [inline addons/web/static/src/views/kanban/kanban_controller.js:127]:
```suggestion
    get progressBarAggregates() {
        const res = []; 
        const { progressAttributes } = this.props.archInfo;
        if (!progressAttributes && progressAttributes.sumField) {
            res.push(progressAttributes.sumField);
            }
            return res;
    }

```

File: addons/crm/static/src/views/crm_kanban/crm_column_progress.js [hunk 2]
```diff
@@ -18,20 +18,19 @@ export class CrmColumnProgress extends ColumnProgress {
...
             }
         });
     }
 
     getRecurringRevenueGroupAggregate(group) {
-        const rrField = this.props.progressAttributes.recurring_revenue_sum_field;
-        const value = group.getAggregates(rrField.name);
-        const title = rrField.string || this.env._t("Count");
+        const rrField = this.props.progressBarState.progressAttributes.recurring_revenue_sum_field;
+        const aggregatedValues = this.props.progressBarState.getAggregateValue(group, rrField);
```
Comment (aab-odoo) [inline addons/crm/static/src/views/crm_kanban/crm_column_progress.js:29]:
```suggestion
        const aggregatedValue = this.props.progressBarState.getAggregateValue(group, rrField);
```

File: addons/web/static/src/views/kanban/kanban_header.js [hunk 3]
```diff
@@ -121,13 +125,18 @@ export class KanbanHeader extends Component {
             body: this.env._t(
                 "Are you sure that you want to archive all the records from this column?"
             ),
-            confirm: () => this.group.list.archive(),
+            confirm: async () => {
+                await this.group.list.archive();
+                this.props.progressBarState?.updateCountsArchive(this.group);
+                // this.props.progressBarState?.updateCounts();
```
Comment (aab-odoo) [inline addons/web/static/src/views/kanban/kanban_header.js:131]: keep updateCounts

File: addons/web/static/src/views/kanban/kanban_header.js [hunk 4]
```diff
@@ -121,13 +125,18 @@ export class KanbanHeader extends Component {
...
+            },
             cancel: () => {},
         });
     }
 
     unarchiveGroup() {
         this.group.list.unarchive();
+        // this.props.progressBarState?.updateCounts();
```
Comment (aab-odoo) [inline addons/web/static/src/views/kanban/kanban_header.js:139]: check, but probably keep

File: addons/web/static/src/views/kanban/kanban_renderer.js [hunk 5]
```diff
@@ -460,6 +466,7 @@ export class KanbanRenderer extends Component {
             const refId = previous ? previous.dataset.id : null;
             const targetGroupId = parent?.dataset.id;
             await this.props.list.moveRecord(dataRecordId, dataGroupId, refId, targetGroupId);
+            // this.props.progressBarState?.updateCountsMove(dataRecordId, dataGroupId, targetGroupId);
```
Comment (aab-odoo) [inline addons/web/static/src/views/kanban/kanban_renderer.js:469]: check if necessary, but probably keep (check rpcs)

File: addons/web/static/src/views/kanban/progress_bar_hook.js [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+                bars.find((b) => b.value === this.activeBars[group.value].value).count === 0
+            ) {
+                group.applyFilter(undefined).then(() => {
+                    delete this.activeBars[group.value];
+                    group.model.notify();
+                });
+            }
+
+            const progressBar = {
+                __activeBars: this.activeBars,
```
Comment (aab-odoo) [inline addons/web/static/src/views/kanban/progress_bar_hook.js:93]: alternative const self = this;

Comment (aab-odoo) [reply addons/web/static/src/views/kanban/progress_bar_hook.js:93]: other alternative (to check)
const activeBars = this.activeBars;

File: addons/web/static/src/views/kanban/progress_bar_hook.js [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+        this.progressAttributes = progressAttributes;
+        this.model = model;
+        this._groupsInfo = {};
+        this._summedFields = summedFields;
+        this.activeBars = activeBars;
+        this._sumAggregates = [];
+        this._pbCounts = null;
+    }
+
+    getGroupInfo(groupId) {
```
Comment (aab-odoo) [inline addons/web/static/src/views/kanban/progress_bar_hook.js:42]: uniformize apis (always give group or groupId) ?

File: addons/web/static/src/views/kanban/progress_bar_hook.js [hunk 8]
```diff
@@ -0,0 +1,346 @@
+/** @odoo-module **/
+
+import { onWillStart, onWillUpdateProps, reactive, useComponent } from "@odoo/owl";
+import { Domain } from "@web/core/domain";
+import { _t } from "@web/core/l10n/translation";
+const FALSE = Symbol("False");
+
+function _findElementInArrayByGroup(array, group) {
```
Comment (aab-odoo) [inline addons/web/static/src/views/kanban/progress_bar_hook.js:8]:
```suggestion
groups: groups returned by web_read_group
function _findGroup(groups, groupByField, value) {
}
```

File: addons/web/static/src/views/kanban/progress_bar_hook.js [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+    }
+    return filterDomain;
+}
+
+class ProgressBarState {
+    constructor(progressAttributes, model, summedFields, activeBars = {}) {
+        this.progressAttributes = progressAttributes;
+        this.model = model;
+        this._groupsInfo = {};
+        this._summedFields = summedFields;
```
Comment (aab-odoo) [inline addons/web/static/src/views/kanban/progress_bar_hook.js:36]: aggregateFields

File: addons/web/static/src/views/kanban/progress_bar_hook.js [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+        this.activeBars = activeBars;
+        this._sumAggregates = [];
```
Comment (aab-odoo) [inline addons/web/static/src/views/kanban/progress_bar_hook.js:38]: aggregateValues

File: addons/web/static/src/views/kanban/progress_bar_hook.js [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+                    const resGroup = _findElementInArrayByGroup(res.groups, group);
+                    activeBar.aggregates = {
+                        ...resGroup,
+                        [group.groupByField.name]: group.value,
+                    };
+                }
+            });
+    }
+
+    async updateCounts(record) {
```
Comment (aab-odoo) [inline addons/web/static/src/views/kanban/progress_bar_hook.js:189]: not async

Comment (aab-odoo) [reply addons/web/static/src/views/kanban/progress_bar_hook.js:189]:
```suggestion
    async updateCounts(group) {
```

Comment (aab-odoo) [reply addons/web/static/src/views/kanban/progress_bar_hook.js:189]:
```suggestion
    async updateCounts(group) {
```

File: addons/web/static/src/views/kanban/progress_bar_hook.js [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+            const domain = this.model.root.domain;
+            const context = this.model.root.context;
+            const { colors, fieldName: field, help } = this.progressAttributes;
+            const res = await this.model.orm.call(resModel, "read_progress_bar", [], {
+                domain,
+                group_by: groupBy.length ? groupBy[0] : defaultGroupBy,
+                progress_bar: { colors, field, help },
+                context,
+            });
+            for (const group of this.model.root.groups) {
```
Comment (aab-odoo) [inline addons/web/static/src/views/kanban/progress_bar_hook.js:268]: be careful here, the groups could have been reloaded meanwhile

**Review by aab-odoo — APPROVED** [review]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo rebase-ff

**Pull request closed without merging**